src/WalkingModule/src/Module.cpp

```diff
@@ -1314,3 +1641,289 @@ bool WalkingModule::stopWalking()
+
+    iDynTree::Position tempPosition;
+    iDynTree::LinVelocity tempVel;
+    iDynTree::toEigen(tempPosition)=iDynTree::toEigen(desiredFootTransform.getPosition())+m_kFootSmoother*(iDynTree::toEigen(adaptedFeetTransform.getPosition())-iDynTree::toEigen(desiredFootTransform.getPosition()));
+    iDynTree::toEigen(tempVel)=iDynTree::toEigen(desiredFootTwist.getLinearVec3())+m_kFootSmoother*(iDynTree::toEigen(adaptedFeetTwist.getLinearVec3())-iDynTree::toEigen(desiredFootTwist.getLinearVec3()));
+    smoothedFootTransform.setPosition(tempPosition);
+    smoothedFootTwist.setLinearVec3(tempVel);
+    return true;
+}
+
+bool WalkingModule::dcmSmoother(const iDynTree::Vector2 adaptedDCM,const iDynTree::Vector2 desiredDCM,iDynTree::Vector2& smoothedDCM )
+{
+    if(m_pushRecoveryActiveIndex<(m_stepAdapter->getPushRecoveryActivationIndex()+1) )
+    {
+        m_kDCMSmoother=0;
+        m_timeIndexAfterPushDetection=0;
+    }
+    else if (m_pushRecoveryActiveIndex>=(m_stepAdapter->getPushRecoveryActivationIndex()+1) && m_timeIndexAfterPushDetection<(m_indexSmoother))
+    {
+        m_timeIndexAfterPushDetection=m_timeIndexAfterPushDetection+1;
+
+        m_kDCMSmoother=((double)(m_timeIndexAfterPushDetection)/((double)(m_indexSmoother)));
+        if(m_timeIndexAfterPushDetection==m_indexSmoother)
+        {
+            m_indexSmoother=0;
+            m_pushRecoveryActiveIndex=0;
+        }
+    }
+    else {
+        m_kDCMSmoother=1;
+    }
+
+    iDynTree::toEigen(smoothedDCM)=iDynTree::toEigen(desiredDCM)+m_kDCMSmoother*(iDynTree::toEigen(adaptedDCM)-iDynTree::toEigen(desiredDCM));
+    return true;
+}
+
+bool WalkingModule::runStepAdaptation(iDynTree::Vector2 measuredZMP)
+{
+    if (m_useStepAdaptation)
+    {
+        m_isPitchActive=0;
+        m_isRollActive=0;
+        m_stepAdapter->triggerStepAdapterByArmCompliant(m_robotControlHelper->getActuatedDoFs(),m_qDesired,m_robotControlHelper->getJointPosition(),
+                                                        m_leftInContact,m_rightInContact,m_robotControlHelper->getAxesList());
+        m_isRollActive=m_stepAdapter->isArmRollActive();
+        m_isPitchActive=m_stepAdapter->isArmPitchActive();
+        if(!m_stepAdapter->UpdateDCMEstimator(m_stableDCMModel->getCoMPosition(),m_stableDCMModel->getCoMVelocity(),measuredZMP,m_comHeightTrajectory.front()))
+        {
+            yError() << "[WalkingModule::updateModule] Unable to to recieve DCM from pendulumEstimator";
+            return false;
+        }
+        m_dcmEstimatedI= m_stepAdapter->getEstimatedDCM();
+    }
+
+    if(m_useStepAdaptation)
+    {
+        //step adjustment
+        double comHeight;
+        double omega;
+        if(!m_trajectoryGenerator->getNominalCoMHeight(comHeight))
+        {
+            yError() << "[updateModule] Unable to get the nominal CoM height!";
+            return false;
+        }
+        omega = sqrt(9.81/comHeight);
+
+        StepAdapterOutput output;
```

Comment:
``output`` requires a lot of memory allocation. I would suggest to store it in the ``stepAdapter`` object and to return it as a const reference of the ``runStepAdaptation``. Since right now ``runStepAdaptation`` it returns a boolean, you can include this boolean in the output. The same would hold also for ``input``. I would save it in the ``Module`` class and avoid to reallocate it every time.

Reply:
I tried to do this but I think it will cause some problem, because the `output` structure includes something that are not only output, but also they are input. It means that `runStepAdaptation` received the values of them and modify them. I didn't get how we can solve this problem?

Reply:
Can you make some examples?

Reply:
For example this value:
https://github.com/MiladShafiee/walking-controllers/blob/8e42cd4b1aedd60192351d26d753847cf57f856f/src/StepAdaptationController/src/StepAdaptationController.cpp#L790

`PushRecoveryActiveIndex` that whenever it becomes 6 it will activate the step adjustment!

Reply:
I moved these structures into the `walkingModule` header.